### PR TITLE
poc - tool for crowdin translation updates

### PR DESCRIPTION
**Pull request by lislei**
@TuxPaper 
I [previously I mentioned](https://github.com/BiglySoftware/BiglyBT/commit/c81885c7e55fe8917efb91ef184731b3e0e456eb#r32921680)
 a possible way of updating changes to the translation files with Crowdin.

I wrote a ugly, nasty, poc that collects and assembles a zip file for uploading to crowdin.
ref I hope you would consider trying it out to see if it's a possible way of automating manual changes to the language files. 

Run with
 `gradlew stageLocalizations -Pcommit=ddaad11a`

Then upload the zip with crowdin.

### CODE REVIEW

**Comment by TuxPaper** [comment]
I can't find your post about requesting more Crowdin access.  Were you able to upload zip files yourself within Crowdin?  I see within Crowdin that you are marked as a Proofreader for the whole BiglyBT project, so I'm hoping that gave you access to the zip upload feature?

**Comment by lislei** [comment]
I found a way to upload the current norwegian translation file once, but that's it. 
When I try the same with a zip file i get the following
bilde

**Comment by TuxPaper** [comment]
hmm, does it allow you to get here? https://crowdin.com/project/biglybt/settings#translations
image

**Comment by TuxPaper** [comment]
Ahh, nm, "Proofreader – does not have access to project setting" from https://support.crowdin.com/modifying-project-participants-roles/

:(

**Comment by TuxPaper** [comment]
"Manager" access level seems overkill and risky (no offense to you!).  API Key has no granularity either.. once one has the API key they can delete the whole project!

Perhaps for now, providing a zip file and having either parg or I click the Upload Translation button will work.  With your "changes only" zip, it should process very fast :)

**Comment by lislei** [comment]
I'm flattered - so no offence taken what so ever.

As a proofreader i get 403 forbidden trying to access those pages. 

Crowdin offers a [REST API for updating files](https://support.crowdin.com/api/update-file/).
Interesting thing there are two ways to authenticate, one is with the api `project-key` that you described. 
The second way is with your personal `account-key` together with your `username`.

Maybe I'll extend this POC and give it a try.

**Comment by lislei** [comment]
Were you able to upload a generated zip to Crowdin btw?

**Comment by TuxPaper** [comment]
Sorry, I've been busy trying to make my ConfigSection stuff ready for committing, so I haven't tried it yet.  We do have Jenkins (Open Source Automation system) internally, that I wouldn't mind integrating what you have into.  Even with my ConfigSection changes, I ran into 1/2 dozen strings I wish I could have did a find/replace on.

**Comment by lislei** [comment]
At my work we're using Jenkins as well. I guess it would be possible to create an automated step for this. 

Are you wanting a manually triggered parameterized job or something automated?

**Comment by lislei** [comment]
(sorry, I accidentaly pressed the close button)

**Comment by lislei** [comment]
I toyed with Crowdins REST API and found a way to upload the translation updates as single files (one by one). I have to admit it took some trail and error on both aspects of Java and Crowdin. 

This call works for me having proofreader access. 
Havning that said, I haven't tested it to see if it behaves as expected. 
I've left the `auto_approve_imported` option off for now - not sure what's the right strategy here.

File: crowdin/build.gradle [hunk 1]
```diff
@@ -106,3 +106,165 @@ task stageLocalizations(type: Zip) {
...
+                'ceb'   : '_ceb_PH.properties',
+                'cs'    : '_cs_CZ.properties',
+                'da'    : '_da_DK.properties',
+                'de'    : '_de_DE.properties',
+                'el'    : '_el_GR.properties',
+                'en-GB' : '_en_GB.properties',
+                'en-US' : '_en_US_classic.properties', //*
+                'es-ES' : '_es_ES.properties',
+                'es-VE' : '_es_VE.properties',
+                'eu'    : '_eu.properties', //not picked up by Java/BiglyBT... guessing it's Basque ??
```
Comment (lislei) [inline crowdin/build.gradle]: I found it to be three languages that aren't picked up by the ui client. They seem to be in Crowdin, but you're not able to choose it under Settings -> Language

**Comment by TuxPaper** [comment]
99% sure we'll need auto_approve_imported on, otherwise it just adds it to CrowdIn but keeps the previous entry approved.

The classic.properties can be set by the user in Interface->Language.  Not sure how _eu works.. I just assumed the people who started the language knew :P  What's the third one?

**Comment by parg** [comment]
_eu / Basque used to work, as did _oc / Occitan. Indeed, _eu is one of our few active translations.

**Comment by parg** [comment]
Just checked - there in 1800, gone in 1900 - same Java version of  1.8_202

**Comment by lislei** [comment]
> 99% sure we'll need auto_approve_imported on, otherwise it just adds it to CrowdIn but keeps the previous entry approved.
> 
I wish the method had the update_option. If the string isn't already approved then changing state doesn't seem right for this usecase.

**Comment by lislei** [comment]
> The classic.properties can be set by the user in Interface->Language. Not sure how _eu works.. I just assumed the people who started the language knew :P What's the third one?

The third one is "Serbian" / `sr`

**Comment by lislei** [comment]
> 
> 
> Just checked - there in 1800, gone in 1900 - same Java version of 1.8_202

I believe it must have been broken during the last refactor. 
I added a unit-test along with a fix in #1022

**Comment by TuxPaper** [comment]
reopening after accidental closing.  Pull request is still of interest.

One additional note is Crowdin now lets us modify a value in the main (english) language file on github, and keep the translations in the other languages (previously CrowdIn would wipe out all translations for the modified key)

**Review by uwxpro — APPROVED** [review]

**Comment by ferdnyc** [comment]
Sorry if I missed a conversation... but, why do files need to be uploaded to Crowdin?

Crowdin is tied in to the project repo, which remains the authoritative data repository. Any changes made there are automatically submitted back here as pull requests. Any _other_ changes that are made _here_ should flow **out** to Crowdin via the repo.

Is there any reason that manual translation edits can't be opened in Github as pull requests, same as any other content changes? There's already a process in place for incorporating changes submitted via Git. And AIUI, Crowdin should be able to (automatically?) pick those up and incorporate them when they're merged.

**Comment by ferdnyc** [comment]
Aha. [Not automatic](https://support.crowdin.com/github-integration/#uploading-translations-from-repo). Buuut, the [GitHub Action](https://support.crowdin.com/github-integration/#uploading-translations-from-repo) can be used to automate uploads from the repo to Crowdin, as a trigger on pull request merges that modify translation files.

**Comment by ferdnyc** [comment]
(An action something like this should do it, say as `.github/workflows/translation-upload.yml`:)
```yml
name: Crowdin upload
on:
  push:
    branches: [ master ]
    paths: core/src/com/biglybt/internat/*
jobs:
  crowdin-upload:
    runs-on: ubuntu-latest
   steps:
     - name: Checkout
       uses: actions/checkout@v2

     - name: crowdin upload
       uses: crowdin/github-action@1.0.4
       with:
         upload_sources: false
         download_translations: false
         upload_translations: true
         auto_approve_imported: true
         translation: 'core/src/com/biglybt/internat/*.properties'
         project_id: ${{ secrets.CROWDIN_PROJECT_ID }}
         token: ${{ secrets.CROWDIN_PERSONAL_TOKEN }}
```

The crowdin project ID (API v2) comes from https://crowdin.com/project/biglybt/settings#api, the crowdin personal token generated at https://crowdin.com/settings#api-key. Both would have to be set as repository secrets at https://github.com/BiglySoftware/BiglyBT/settings/secrets/actions.

**Comment by lislei** [comment]
@ferdnyc you are right. There isn't an automated way to backflow into Crowdin as you might reasonably expect. 

I think github actions are worth considering.
@TuxPaper @parg, if I may, whats your vote on this?

**Comment by lislei** [comment]
@ferdnyc - your suggested github action has my vote. I'd be interested in such a solution - please give me a shout!

Closing this issue.

**Comment by ferdnyc** [comment]
@lislei 

I mean, I can open a PR to add that proposed workflow to the `.github/workflows/` directory in the repo, but either @parg or @TuxPaper would need to create the necessary secrets if it's going to do anything other than fail to authenticate with Crowdin.

But, I guess a PR is a reasonable way of opening that discussion back up. I'll put something together.

**Pull request closed without merging**